Migrate to ESLint

gemmaro opened the pull request:
Fix #28

na2hiro commented on package.json:
```diff
@@ -35,15 +35,18 @@
     "pre-push": "^0.1.1",
     "ts-jest": "^26.4.3",
     "ts-loader": "^4.3.1",
     "tslint": "^5.9.1",
```
Thanks for the pull request! It looks good.

Could you run `npm uninstall tslint` to remove the dependency?

gemmaro replied:
Thank you for your checking. I'll do it this evening.

gemmaro commented:
Current status:

- [ ] `tslint.json` -> `.eslintrc.js`
  - I'm not sure `tslint.json` affects ESLint behaviour
- [ ] Remove TSLint comments from `**/__tests__/*`

```shell-session
$ rg tslint       
src/__tests__/shogiTest.ts
606:            KE: 0, // tslint:disable-line object-literal-sort-keys
619:            KE: 0, // tslint:disable-line object-literal-sort-keys
630:            KE: 0, // tslint:disable-line object-literal-sort-keys

tslint.json
2:  "extends": "tslint:recommended",
```

gemmaro commented:
I'm really sorry, I just realized that I didn't take into account the `.nvmrc` file. (This is because I was using nodenv to version control Node.js.)
So I just made a symbolic link from `.nvmrc` to `.node-version` to see if ESLint would work with the Node.js version in `.nvmrc`, and **it didn't**.
I guess I need to upgrade the Node.js version first.

Secondly, when initializing ESLint, I made ESLint guess how to format the code from the existing code. However, this guessing does not seem to work very well, and is quite sparse, unlike the current style.

There are at least two problems, and if I continue, I'll have to force-commit almost from the beginning again, so I'd like to close this PR once.

na2hiro commented:
@gemmaro We should bump Node version from current 10 because its LTS support has finished. Could you tell me the version of node you used to create this PR? We can update `.nvmrc`, `.travis.yml` and `.github/workflows/gh-pages.yml` (this uses 12 :) ) to match with your version if that would help.

If not I'll update Node version independently (https://github.com/na2hiro/Shogi.js/issues/31)

gemmaro commented:
> Could you tell me the version of node you used to create this PR?

I was using v14.17.5 (`$ node --version`), just for reference.

In another point of view, I visit [the Node.js website](https://nodejs.org/en/) now, and see it recommends v16.13.1.
I'm not familiar with JavaScript ecosystem very much, but v16.x.x seems to be good choice.

Note that I I would like to clean-up my commits of this PR anyway. If another PR determines new Node.js version, I'll merge it to this one 👍

na2hiro commented:
Sounds good! Let me bump Node to 16 then

gemmaro committed:
Add ESLint as dev-depends

gemmaro committed:
Init ESLint config

gemmaro committed:
Configure ESLint and remove TSLint

na2hiro commented:
LGTM

gemmaro commented:
As a reminder, no rules have been applied for styles at this time. Coding conventions might be defined separately.
(In fact, there seems to be a variety of them, but I have no idea which one is the best.)

na2hiro committed:
Enable ESLint for jest tests

na2hiro commented:
Do you mean something like where to put spaces and new lines etc are yet to be determined? I kind of liked prettier for the stubborn formatting. If you want to integrate that here, I appreciate it, but we can merge as it is as well.

Also I'd like to add `eslint-plugin-jest` for linting test cases too. I added a commit on top of your branch (I wonder if that's common in GitHub though?)

gemmaro committed:
Install eslint-config-prettier

gemmaro committed:
Activate Prettier for ESLint

gemmaro committed:
Enable eslint-plugin-jest

gemmaro commented:
> Do you mean something like where to put spaces and new lines etc are yet to be determined?

Yes. Thank you for letting me know.

> I kind of liked prettier for the stubborn formatting. If you want to integrate that here, I appreciate it, but we can merge as it is as well.

Okay. I added `eslint-config-prettier`. (Reference: [Integrating with Linters](https://prettier.io/docs/en/integrating-with-linters.html))

> Also I'd like to add `eslint-plugin-jest` for linting test cases too. I added a commit on top of your branch (I wonder if that's common in GitHub though?)

👍 

Also enabled Jest in ESLint config

> You can install it using `npm install --save-dev eslint-plugin-jest` or `yarn add --dev eslint eslint-plugin-jest` and it can be enabled by adding `{"plugins": ["jest"]}` to your eslint configuration.

(Reference: [`eslint-plugin-jest` – our very own ESLint plugin](https://jestjs.io/blog/2017/02/21/jest-19-immersive-watch-mode-test-platform-improvements#eslint-plugin-jest--our-very-own-eslint-plugin))

na2hiro committed:
Integrate Prettier

na2hiro commented:
Integrated prettier

na2hiro commented:
Thanks for having initiative to this!
